File: tools/opendss/smoke.py
"""Smoke-test the Oddie bridge over each vendored EPRI OpenDSSDirect.dll.

For every revision in `revisions.json` (one subprocess per revision — Oddie
wraps ONE engine per process; loading the same DLL twice aliases the same
engine, so per-process isolation is the only safe mode):

  1. assert the venv pin (`PIN_OPENDSS.txt`: dss-python / dss-python-backend);
  2. load the DLL via `dss.IOddieDSS(library_path=<abs>)`, print `DSS.Version`;
  3. compile + solve the IEEE13 master from the vendored corpus (inside the
     oracle server's `_CorpusGuard` so no report files pollute the corpus);
  4. prove the CSC Y export is solution-neutral: EPRI's `InitAndGetYparams`
     ALWAYS factors before exporting (DYMatrix.pas), so `YNodeVarray` must be
     bit-identical before/after `getYSparse()` — this is the empirical proof
     the oracle captures can use `full_csc` on the EPRI engine;
  5. sanity-check `getI()` length (= 2*(NumNodes+1), slot 0 = ground);
  6. check `DSS.Version` against `revisions.json` `expect_version`; while the
     pin is empty, print the discovered string as `PIN THIS:` (fill it in and
     re-run — the oracle server refuses an empty pin, never a silent pass).

Run:  tools/opendss/.venv/Scripts/python tools/opendss/smoke.py
"""

from __future__ import annotations

import json
import os
import subprocess
import sys
from pathlib import Path

HERE = Path(__file__).resolve().parent
REPO_ROOT = HERE.parents[1]
REVISIONS_JSON = HERE / "revisions.json"
IEEE13 = (
    REPO_ROOT
    / "tests/corpus/electricdss-tst/Version8/Distrib/IEEETestCases/13Bus/IEEE13Nodeckt.dss"
)

sys.path.insert(0, str(REPO_ROOT / "tools" / "oracle"))


def read_pin() -> dict[str, str]:
    pins: dict[str, str] = {}
    for line in (HERE / "PIN_OPENDSS.txt").read_text().splitlines():
        line = line.split("#", 1)[0].strip()
        if "==" in line:
            k, v = line.split("==", 1)
            pins[k.strip()] = v.strip()
    return pins


def check_pin() -> None:
    pins = read_pin()
    import dss
    import dss_python_backend

    if dss.__version__ != pins["dss-python"]:
        sys.exit(
            f"dss-python {dss.__version__} != pinned {pins['dss-python']} "
            "(tools/opendss/PIN_OPENDSS.txt)"
        )
    if dss_python_backend.__version__ != pins["dss-python-backend"]:
        sys.exit(
            f"dss-python-backend {dss_python_backend.__version__} != pinned "
            f"{pins['dss-python-backend']} (tools/opendss/PIN_OPENDSS.txt)"
        )


def run_rev(rev: str) -> None:
    check_pin()
    revs = json.loads(REVISIONS_JSON.read_text())
    spec = revs[rev]
    dll = (REPO_ROOT / spec["dll"]).resolve()
    if not dll.is_file():
        sys.exit(f"{rev}: DLL not vendored: {dll} (run tools/opendss/vendor_binaries.py)")
    if not IEEE13.is_file():
        sys.exit(f"IEEE13 master not found: {IEEE13}")

    from dss import IOddieDSS

    from oracle_server import _CorpusGuard  # noqa: E402  (tools/oracle on sys.path)

    cwd0 = os.getcwd()  # the EPRI engine chdirs the process on Compile
    d = IOddieDSS(library_path=str(dll))
    ver = str(d.Version)
    print(f"{rev}: engine Version = {ver!r}")

    d.AllowForms = False
    try:
        d.AllowEditor = False  # not implemented in Oddie -> DSSException, harmless
    except Exception:
        pass

    try:
        with _CorpusGuard(str(IEEE13)):
            d.Text.Command = "clear"
            d.Text.Command = f'Compile "{IEEE13}"'
            d.Text.Command = "solve"
            ckt = d.ActiveCircuit
            sol = ckt.Solution
            assert sol.Converged, f"{rev}: IEEE13 did not converge"
            n_nodes = len(ckt.YNodeOrder)
            print(f"{rev}: IEEE13 solved, {n_nodes} nodes, {sol.Iterations} iterations")

            # 4. factored-export neutrality: bit-identical voltages around the
            # CSC export (EPRI InitAndGetYparams always factors first).
            v0 = list(ckt.YNodeVarray)
            y = d.YMatrix.getYSparse()
            assert y is not None, f"{rev}: getYSparse returned None"
            data, row_idx, col_ptr = y
            assert len(col_ptr) - 1 == n_nodes, f"{rev}: CSC n={len(col_ptr) - 1} != {n_nodes}"
            assert len(data) > 0, f"{rev}: CSC export empty"
            v1 = list(ckt.YNodeVarray)
            assert v0 == v1, (
                f"{rev}: YNodeVarray CHANGED across getYSparse() — the factored "
                "export is NOT solution-neutral on this engine; use full_csc=false"
            )
            print(f"{rev}: getYSparse OK (n={n_nodes}, nnz={len(data)}), voltages bit-identical")

            # 5. injection vector shape.
            i_vec = list(d.YMatrix.getI())
            assert len(i_vec) == 2 * (n_nodes + 1), (
                f"{rev}: getI length {len(i_vec)} != {2 * (n_nodes + 1)}"
            )
            print(f"{rev}: getI OK (len={len(i_vec)})")
    finally:
        os.chdir(cwd0)

    expect = spec.get("expect_version", "")
    if not expect:
        print(f"{rev}: PIN THIS: expect_version -> {ver!r} (fill revisions.json and re-run)")
    elif expect not in ver:
        sys.exit(f"{rev}: engine {ver!r} does not contain pinned {expect!r}")
    else:
        print(f"{rev}: version pin OK ({expect!r})")
    print(f"{rev}: SMOKE OK")


def main() -> None:
    if "--rev" in sys.argv:
        run_rev(sys.argv[sys.argv.index("--rev") + 1])
        return
    revs = json.loads(REVISIONS_JSON.read_text())
    failed = []
    for rev in revs:
        print(f"=== {rev} ===", flush=True)
        r = subprocess.run([sys.executable, __file__, "--rev", rev])
        if r.returncode != 0:
            failed.append(rev)
    if failed:
        sys.exit(f"smoke FAILED for: {', '.join(failed)}")
    print("=== all revisions OK ===")


if __name__ == "__main__":
    main()

File: crates/dss-epri/src/smoke.rs
//! Oracle-free self-smoke for the r4133 bridge (`UNIFIED_GATE_PLAN.md` §2.4-1,
//! replacing `tools/opendss/smoke.py`): DLL loads, version matches
//! `revisions.json`, IEEE13 compiles + solves + converges, the CSC export is
//! solution-neutral (`YNodeVarray` bit-identical before/after
//! `InitAndGetYparams`/`GetCompressedYMatrix`), and the injection vector has the
//! `2*(NumNodes+1)` shape. Shared by the `epri-worker --smoke` mode and the
//! `smoke_*` integration `#[test]` so `cargo test --workspace` exercises it with
//! no oracle installed.

use std::path::PathBuf;

use crate::dss::{Engine, EngineError};

/// The vendored r4133 DLL. `DSS_EPRI_DLL` overrides; default is relative to the
/// crate (git-tracked at `tools/opendss/bin/r4133/OpenDSSDirect.dll`).
pub fn dll_path() -> PathBuf {
    if let Ok(p) = std::env::var("DSS_EPRI_DLL") {
        return PathBuf::from(p);
    }
    workspace_root().join("tools/opendss/bin/r4133/OpenDSSDirect.dll")
}

fn workspace_root() -> PathBuf {
    // CARGO_MANIFEST_DIR = <root>/crates/dss-epri (baked at build time).
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("..")
        .join("..")
}

/// The r4133 `expect_version` substring: `DSS_EPRI_EXPECT` overrides, else read
/// from `tools/opendss/revisions.json`.
pub fn expect_version() -> Result<String, String> {
    if let Ok(v) = std::env::var("DSS_EPRI_EXPECT") {
        return Ok(v);
    }
    let p = workspace_root().join("tools/opendss/revisions.json");
    let text = std::fs::read_to_string(&p).map_err(|e| format!("read {}: {e}", p.display()))?;
    let v: serde_json::Value =
        serde_json::from_str(&text).map_err(|e| format!("parse {}: {e}", p.display()))?;
    v.get("r4133")
        .and_then(|r| r.get("expect_version"))
        .and_then(|s| s.as_str())
        .filter(|s| !s.is_empty())
        .map(|s| s.to_string())
        .ok_or_else(|| format!("r4133.expect_version missing/empty in {}", p.display()))
}

fn ieee13_case() -> String {
    workspace_root()
        .join("tests/corpus/electricdss-tst/Version8/Distrib/IEEETestCases/13Bus/IEEE13Nodeckt.dss")
        .to_string_lossy()
        .replace('\\', "/")
}

/// The printable smoke report (lines committed to STATUS).
pub struct SmokeReport {
    pub lines: Vec<String>,
}

/// Run the smoke checks against the r4133 DLL. Returns the report on success,
/// an error string on any failed check.
pub fn run_smoke() -> Result<SmokeReport, String> {
    let dll = dll_path();
    if !dll.is_file() {
        return Err(format!("r4133 DLL not found: {}", dll.display()));
    }
    let engine = Engine::new(&dll).map_err(|e: EngineError| e.to_string())?;
    let mut lines = Vec::new();

    // 1. version pin.
    let ver = engine.version().trim().to_string();
    let expect = expect_version()?;
    if !ver.contains(&expect) {
        return Err(format!(
            "engine version {ver:?} does not contain {expect:?}"
        ));
    }
    lines.push(format!("version OK: {ver}"));

    // 2. IEEE13 compile + solve converges.
    let ieee13 = ieee13_case();
    if !PathBuf::from(&ieee13).is_file() {
        return Err(format!("IEEE13 master missing: {ieee13}"));
    }
    engine.clear().map_err(|e| e.to_string())?;
    engine.compile(&ieee13, false).map_err(|e| e.to_string())?;
    engine.solve(false).map_err(|e| e.to_string())?;
    if !engine.converged() {
        return Err("IEEE13 did not converge".to_string());
    }
    let n_nodes = engine.num_nodes();
    let iters = engine.iterations();
    lines.push(format!(
        "IEEE13 solved: {n_nodes} nodes, {iters} iterations"
    ));

    // 3. CSC export solution-neutral (InitAndGetYparams always factors first).
    let v0 = engine.ynode_varray();
    let ycsc = engine.y_csc().map_err(|e| e.to_string())?;
    if ycsc.n != n_nodes as usize {
        return Err(format!("CSC n={} != NumNodes {n_nodes}", ycsc.n));
    }
    let nnz = ycsc.row_idx.len();
    if nnz == 0 {
        return Err("CSC export empty".to_string());
    }
    let v1 = engine.ynode_varray();
    if v0 != v1 {
        return Err("YNodeVarray CHANGED across the CSC export — not solution-neutral".to_string());
    }
    lines.push(format!(
        "CSC export OK: n={n_nodes}, nnz={nnz}, voltages bit-identical (solution-neutral)"
    ));

    // 4. injection vector shape (getIpointer readable, length 2*(NumNodes+1)).
    let inj = engine.injection_raw(n_nodes);
    let want = 2 * (n_nodes as usize + 1);
    if inj.len() != want {
        return Err(format!("getIpointer length {} != {want}", inj.len()));
    }
    lines.push(format!(
        "getIpointer OK: len={} (= 2*(NumNodes+1))",
        inj.len()
    ));

    // 5. all-properties enumeration round-trips (§2.2 report-tooling parity):
    // `DSSElementV` (AllPropertyNames) + `? name.prop` value reads. Re-reading the
    // first element's first property directly must equal the dumped value.
    let dump = crate::capture::all_properties_dump(&engine).map_err(|e| e.to_string())?;
    if dump.is_empty() {
        return Err("all_properties dump is empty (no elements enumerated)".to_string());
    }
    let first = &dump[0];
    let (p0, v0) = first
        .props
        .first()
        .ok_or_else(|| format!("all_properties: {} has no properties", first.element))?;
    let direct = engine.raw_command(&format!("? {}.{}", first.element, p0));
    if &direct != v0 {
        return Err(format!(
            "all_properties round-trip mismatch on {}.{p0}: dump {v0:?} != direct {direct:?}",
            first.element
        ));
    }
    let total: usize = dump.iter().map(|p| p.props.len()).sum();
    lines.push(format!(
        "all_properties OK: {} elements, {total} property values, round-trip verified",
        dump.len()
    ));

    Ok(SmokeReport { lines })
}
